tests/ipc/StateIPCTestServer.ml

```diff
@@ -0,0 +1,181 @@
+ * before the contract is executed. *)
+
+open Core
+open Result.Let_syntax
+open MonadUtil
+open StateIPCIdl
+
+module Hashtbl = Caml.Hashtbl
+
+type hashtable =  (string, value_type) Hashtbl.t
+and value_type =
+  | NonMapVal of string
+  | MapVal of hashtable
+
+let thread_pool : (string, hashtable) Hashtbl.t =  Hashtbl.create 4
+
+let num_pending_requests = 5
+let fetch_message = "Fetching state value failed"
+let update_message = "Updating state value failed"
+let fail a = Error a
+
+(* Send msg with delimiting character "0xA". *)
+let send_delimited oc msg =
+  let msg' = msg ^ "\n" in
+  Out_channel.output_string oc msg';
+  Out_channel.flush oc
+
+let decode_serialized_value value =
+  let decoder = Pbrt.Decoder.of_bytes (Bytes.of_string value) in
+  Ipcmessage_pb.decode_proto_scilla_val decoder
+
+let encode_serialized_value value =
+  let encoder = Pbrt.Encoder.create () in
+  Ipcmessage_pb.encode_proto_scilla_val value encoder;
+  Bytes.to_string @@ Pbrt.Encoder.to_bytes encoder
+
+let decode_serialized_query query =
+  let decoder = Pbrt.Decoder.of_bytes (Bytes.of_string query) in
+  Ipcmessage_pb.decode_proto_scilla_query decoder
+
+module MakeServer() = struct
+
+  module IPCTestServer = IPCIdl(Idl.GenServer ())
+
+  (* Global state of the server thread. *)
+  let table = Hashtbl.create 8
+
+  let binary_rpc conn =
+    let ic = Unix.in_channel_of_descr conn in
+    let oc = Unix.out_channel_of_descr conn in 
+    let request = Jsonrpc.call_of_string (Caml.input_line ic) in
+    let response = (Idl.server IPCTestServer.implementation) request in
+    send_delimited oc (Jsonrpc.string_of_response response)
+
+  let prepare_server sock_addr =
+    (try Unix.unlink sock_addr with Unix.Unix_error(Unix.ENOENT, _, _) -> ());
+    let socket = Unix.socket ~domain: Unix.PF_UNIX ~kind: Unix.SOCK_STREAM ~protocol:0 in
+    Unix.bind socket ~addr:(Unix.ADDR_UNIX sock_addr);
+    Unix.listen socket ~backlog:num_pending_requests;
+    let server () =
+      while true do
+        let conn, _ = Unix.accept socket in
+        binary_rpc conn;
+        Unix.close conn;
+      done
+    in
+    server
+
+  let rec serialize_value value =
+    match value with
+    | NonMapVal v -> Ipcmessage_types.Bval (Bytes.of_string v)
+    | MapVal m -> 
+      let map_list = Hashtbl.fold (fun k v acc -> (k, v) :: acc) m [] in
+      let serialized_map_list = List.map map_list ~f: (fun (str, v) -> (str, serialize_value v)) in
+      Ipcmessage_types.Mval({ m = serialized_map_list})
+
+  let rec deserialize_value value =
+    match value with 
+    | Ipcmessage_types.Bval v -> NonMapVal (Bytes.to_string v)
+    | Ipcmessage_types.Mval m_list ->
+      let new_table = Hashtbl.create (List.length m_list.m) in
+      List.iter m_list.m ~f: (fun (str, v) -> 
+        Hashtbl.replace new_table str (deserialize_value v));
+      MapVal new_table
+
+  (* Not sure if need to check mapdepth vs length of indices? *)
```

Comment:
What would be the point? If too many indices are supplied, then the fetch will fail. If too few are supplied, then that's indistinguishable from fetching a submap.

Reply:
Sorry, the comment was (I think) added when Advay initially started writing this file, but I later missed to remove it. Removed it now.